apps/emqx_ds_backends/test/emqx_ds_backends_SUITE.erl

```diff
@@ -978,6 +978,442 @@ t_sub_catchup_unrecoverable(Config) ->
+    Topic1 = [<<"foo">>],
+    Topic2 = [<<"bar">>],
+    ?check_trace(
+        begin
+            %% Open the database
+            Opts = maps:merge(opts(Config), #{
+                store_ttv => true,
+                storage => {emqx_ds_storage_skipstream_lts_v2, #{}}
+            }),
+            ?assertMatch(
+                ok,
+                emqx_ds_open_db(DB, Opts)
+            ),
+            timer:sleep(100),
+            %% 1. Insert test data:
+            {ok, Tx1} = emqx_ds:new_kv_tx(DB, TXOpts),
+            Ops1 = #{
+                ?ds_tx_write => [{Topic1, 0, <<"payload0">>}]
+            },
+            ?assertMatch({ok, _}, do_commit_tx(DB, Tx1, Ops1)),
+
+            %% 2. Preconditions, successful.
+            {ok, Tx2} = emqx_ds:new_kv_tx(DB, TXOpts),
+            Ops2 = #{
+                ?ds_tx_expected => [
+                    {Topic1, 0, '_'},
+                    {Topic1, 0, <<"payload0">>}
+                ],
+                ?ds_tx_unexpected => [{Topic2, 0}],
+                ?ds_tx_write => [{Topic1, 0, <<"payload1">>}]
+            },
+            ?assertMatch(
+                {ok, _},
+                do_commit_tx(DB, Tx2, Ops2)
+            ),
+            ?assertEqual(
+                [{Topic1, 0, <<"payload1">>}],
+                emqx_ds:dirty_read(DB, ['#'])
+            ),
+
+            %% 3. Precondiction fail, unexpected message:
+            {ok, Tx3} = emqx_ds:new_kv_tx(DB, TXOpts),
+            Ops3 = #{
+                ?ds_tx_unexpected => [{Topic1, 0}],
+                ?ds_tx_write => [{Topic1, 0, <<"fail">>}]
+            },
+            ?assertMatch(
+                ?err_unrec({precondition_failed, _}),
+                do_commit_tx(DB, Tx3, Ops3)
+            ),
+
+            %% 4 Preconditions, fail, expected message not found:
+            {ok, Tx4} = emqx_ds:new_kv_tx(DB, TXOpts),
+            Ops4 = #{
+                ?ds_tx_expected => [{Topic2, 0, '_'}],
+                ?ds_tx_write => [{Topic2, 0, <<"fail">>}]
+            },
+            ?assertMatch(
+                ?err_unrec({precondition_failed, _}),
+                do_commit_tx(DB, Tx4, Ops4)
+            ),
+
+            %% 5 Preconditions, fail, value is different:
+            {ok, Tx5} = emqx_ds:new_kv_tx(DB, TXOpts),
+            Ops5 = #{
+                ?ds_tx_expected => [{Topic1, 0, <<"payload0">>}],
+                ?ds_tx_write => [{Topic1, 0, <<"fail">>}]
+            },
+            ?assertMatch(
+                ?err_unrec(
+                    {precondition_failed, [
+                        #{expected := <<"payload0">>, got := <<"payload1">>, topic := Topic1}
+                    ]}
+                ),
+                do_commit_tx(DB, Tx5, Ops5)
+            ),
+
+            %% Verify that side effects of failed transactions weren't
+            %% applied:
+            ?assertEqual(
+                [{Topic1, 0, <<"payload1">>}],
+                emqx_ds:dirty_read(DB, ['#'])
```

Comment:
I think it's better to use transaction read here to rule out bad data lingering in the buffer.